spectec/src/il/validation.ml

```diff
@@ -241,7 +241,15 @@ let valid_list valid_x_y env xs ys at =
 let rec valid_iter env iter =
   match iter with
   | Opt | List | List1 -> ()
-  | ListN (e, _) -> valid_exp env e (NatT $ e.at)
+  | ListN (e, None) -> valid_exp env e (NatT $ e.at)
+  | ListN (e, Some id) ->
+    valid_exp env e (NatT $ e.at);
+    let t', dim = find "variable" env.vars id in
+    equiv_typ env t' (NatT $ e.at) e.at;
+    if dim <> [ListN (e, None)] then
+      error e.at ("use of iterated variable `" ^
+        id.it ^ String.concat "" (List.map string_of_iter dim) ^
+        "` outside suitable iteraton context")
```

Comment:
I'm afraid plain equality does not work for this, because terms usually differ by region. To compare, you need to use the functions from the eq.ml module, which ignore region annotations.
```suggestion
    match dim with
    | [ListN (e', None)] when Eq.eq_exp e e' -> ()
    | _ ->
      error e.at ("use of iterated variable `" ^
        id.it ^ String.concat "" (List.map string_of_iter dim) ^
        "` outside suitable iteraton context")
```

Reply:
Or actually this, if you add eq_list to the exports of eq.ml:
```suggestion
    if not Eq.(eq_list eq_iter dim [ListN (e, None)]) then
      error e.at ("use of iterated variable `" ^
        id.it ^ String.concat "" (List.map string_of_iter dim) ^
        "` outside suitable iteraton context")
```

Reply:
Oh, I overlooked it. Thank you!